ddht/v5/crawl.py

```diff
@@ -0,0 +1,280 @@
+            self.bad_enr_count += 1
+        except UnexpectedMessage as error:
+            # TODO: Nodes sometimes send us Nodes messages with an unexpectedly large
+            # number of peers. 12 or 15 of them. We should probably accept these messages!
+            logger.exception("[{task}] Received a bad message from the peer.  nodeid={encode_hex(remote_enr.node_id)} enr={remote_enr}")
+            self.bad_enr_count += 1
+
+    async def read_from_queue_until_done(self):
+        task = get_current_task()
+        async for enr in self.enrqueue:
+            await self.visit_enr(enr)
+        await self.manager.stop()
+
+    async def schedule_enr_to_be_visited(self, enr):
+        task = get_current_task()
+
+        if enr.node_id in self.seen_nodeids:
+            return
+
+        if IP_V4_ADDRESS_ENR_KEY not in enr:
+            logger.info(f"[{task}] Dropping ENR without IP address enr={enr} kv={enr.items()}")
+            return
+
+        self.seen_nodeids.add(enr.node_id)
+
+        try:
+            self.enr_db.set_enr(enr)
+        except OldSequenceNumber:
+            logger.info(f"[{task}] Dropping old ENR. enr={enr} kv={enr.items()}")
+            return
+
+        logger.info(f"[{task}] Found ENR. count={len(self.seen_nodeids)} enr={enr} kv={enr.items()}")
+
+        await self.enrqueue.send(enr)
+
+    async def run(self) -> None:
+        logger.info("Crawling!")
+
+        boot_info = self._boot_info
+
+        # TODO: use these...
+        boot_info.port: int
+        boot_info.bootnodes: Tuple[ENRAPI]
+        boot_info.private_key: Optional[keys.PrivateKey]
+        boot_info.listen_on
+
+        # TODO: support UPNP?
+
+        # 1. Queue up some ENRs to be crawled
+
+        to_enr = lambda bootnode: ENR.from_repr(bootnode, default_identity_scheme_registry)
+        bootnodes = [to_enr(bootnode) for bootnode in DEFAULT_BOOTNODES[2:3]]
+
+        for bootnode in bootnodes:
+            await self.schedule_enr_to_be_visited(bootnode)
+
+        listen_on = boot_info.listen_on or DEFAULT_LISTEN
+        logger.info(f"About to listen. bind={listen_on}:{boot_info.port}")
+        await self.sock.bind(("0.0.0.0", boot_info.port))
+
+        with self.sock:
+            for service in self.services:
+                self.manager.run_daemon_child_service(service)
+            for _ in range(self.concurrency):
+                self.manager.run_daemon_task(self.read_from_queue_until_done)
+            await self.manager.wait_finished()
+
+        successes = len(self.seen_nodeids) - self.bad_enr_count
+        logger.info(f"Finished crawling. found_enrs={len(self.seen_nodeids)} bad_enrs={self.bad_enr_count} successful_connects={successes}")
+
+
+class ENRQueue:
+    """
+    The script should quit when there is nothing left to crawl.
+
+    However, it can't quit when the ENR queue is empty, because the ENR queue might
```

Comment:
Thanks for the prod to find something better. I had originally decided against doing that because I thought there would be race-conditions. However, we're running in a single thread! There are now some ugly `CAUTION` blocks to prevent future us from adding race-conditions back in but I think you're right that this approach is overall much cleaner.